### PR TITLE
Add spec conformance tests for string template expression

### PR DESCRIPTION
**Pull request by chiranSachintha**

### CODE REVIEW

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 1]
```diff
@@ -0,0 +1,323 @@
+Test-Case: output
+Description: Test whether string-template-expr interpolates the results of evaluating expressions into a literal string.
+Labels: string, function-defn, int, string-template-expr, string-literal
```
Comment (KRVPerera) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:3]: Need `additive-expr`

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+function errorFunction() {
+    string a1 = "a1";
+    int _ = string `${a1}`; // @error expected 'int' but found 'string'
+    int a2 = 12;
+    int _ = string `${a2}`; // @error expected 'int' but found 'string'
+    int a3 = 13;
+    int _ = string `${a2 + a3}`; // @error expected 'int' but found 'string'
```
Comment (KRVPerera) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:29]: Need `additive-expr`

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    string temp7 = string `${a7}`;
+    io:println(temp7); // @output true
+    string:Char a8 = "C";
+    string temp8 = string `${a8}`;
+    io:println(temp8); // @output C
+}
+
+Test-Case: output
+Description: Test type of the expression in each interpolation of string-template-expr.
+Labels: string, function-defn, string-template-expr, int, union-type, float, decimal
```
Comment (KRVPerera) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:66]: Need `boolean`

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    string temp2 = string `${a2}`;
+    io:println(temp2); // @output 12
+    string|int|decimal|float|boolean & readonly a3 = true;
+    string temp3 = string `${a3}`;
+    io:println(temp3); // @output true
+}
+
+Test-Case: output
+Description: Test type of the expression in each interpolation of string-template-expr.
+Labels: string, function-defn, string-template-expr
```
Comment (KRVPerera) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:104]: Need `module-type-defn`, `union-type`, `boolean-literal`, `string-literal`

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    string _ = string `${a6}`; // @error expected '(int|float|decimal|string|boolean)', found '[int,int]'
+}
+
+function f1() returns int {
+    return 10;
+}
+
+Test-Case: error
+Description: Test type of the expression in each interpolation of string-template-expr.
+Labels: string, function-defn, string-template-expr, union-type
```
Comment (KRVPerera) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:186]: Need `map-type`, `boolean-literal`, `array-type`, int

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    string _ = string `${`;
+}
+
+
+Test-Case: error
+Description: Test every character that is not part of an interpolation is interpreted as a literal character.
+Labels: string, function-defn, string-template-expr
+
+function errorFunction() {
+    string _ = string `${`}`; @error backtick as a expression
```
Comment (KRVPerera) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:264]:
```suggestion
    string _ = string `${`}`; // @error backtick as a expression
```

Comment (KRVPerera) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt:264]: I am getting many parser errors here.

**Review by KRVPerera — COMMENTED** [review]
We don't need to add `function-defn` if we are not testing it specially

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 1 above)
Comment (SasinduDilshara) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:3]: Do we need `function-defn` label in here. We do not define a new function inside the test-case.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+    io:println(temp7); // @output 32
+    byte|int:Unsigned8|int:Signed8|int:Unsigned16|int:Signed16|int:Unsigned32 a8 = 23;
+    string temp8 = string `${a8}`;
+    io:println(temp8); // @output 23
+}
+
+Test-Case: error
+Description: Test type of the expression in each interpolation of string-template-expr.
+Labels: string, string-template-expr, int, record-type, optional-type, tuple-type, function-type
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:164]: Don't we need `function-defn` label.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 5 above)
Comment (SasinduDilshara) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:186]: Don't we need `boolean` label

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    io:println(temp6); // @output {}
+}
+
+Test-Case: error
+Description: Test every character that is not part of an interpolation is interpreted as a literal character.
+Fail-Issue: ballerina-platform/ballerina-lang#35515
+Labels: string, function-defn, string-template-expr
+
+function errorFunction() {
+    string _ = string `${`;
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:246]: Missing the error output.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+function f1() {
+    string name = "Ballerina";
+    string _ = string `${name}`;
+}
+
+function f2() {
+    string _ = string `${`;
+}
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:257]: Function `f1` and `f2` are unused in the testcase.

**Comment by SasinduDilshara** [comment]
There are no test-cases for nested string template expressions.
Scenario
```
string x = "abc";
string y = string `${string `${x}`}`;
io:println(y);
```

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+Labels: string, function-defn, int, string-template-expr, string-literal
+
+function errorFunction() {
+    string a1 = "a1";
+    int _ = string `${a1}`; // @error expected 'int' but found 'string'
+    int a2 = 12;
+    int _ = string `${a2}`; // @error expected 'int' but found 'string'
+    int a3 = 13;
+    int _ = string `${a2 + a3}`; // @error expected 'int' but found 'string'
+    int _ = <int> string `${a2 + a3}`; // @error 'string' cannot be cast to 'int'
```
Comment (MaryamZi) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:30]: Let's not combine multiple things in the same test. Missing relevant labels anyway.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+}
+
+Test-Case: output
+Description: Test type of the expression in each interpolation of string-template-expr.
```
Comment (MaryamZi) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:34]: We can improve the description by mentioning the types.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+    string temp5 = string `${a5}`;
+    io:println(temp5); // @output 12
+}
+
+Test-Case: output
+Description: Test type of the expression in each interpolation of string-template-expr.
+Labels: string, function-defn, string-template-expr, int, float, decimal, union-type, intersection-type
+
+function init() {
+    string|int & readonly a1 = "a1";
```
Comment (MaryamZi) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:91]: This is equivalent to `string|int`, so not quite sure if this test adds value.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+    int|float|decimal & readonly a2 = 12;
+    string temp2 = string `${a2}`;
+    io:println(temp2); // @output 12
+    string|int|decimal|float|boolean & readonly a3 = true;
+    string temp3 = string `${a3}`;
+    io:println(temp3); // @output true
+}
+
+Test-Case: output
+Description: Test type of the expression in each interpolation of string-template-expr.
```
Comment (MaryamZi) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:103]: Please improve the descriptions for each test mentioning what exactly is being tested. There are seven tests with exactly this description atm.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 4 above)
Comment (MaryamZi) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt:104]: Also singleton-type.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+    string temp3 = string `$ $ $ ${a2 + a3} $ $ $`;
+    io:println(temp3); // @output $ $ $ 25 $ $ $
+    string temp4 = string `$$${"a2"}$$`;
+    io:println(temp4); // @output $$a2$$
+    string temp5 = string `$v1$${"a2"}$v2$`;
+    io:println(temp5); // @output $v1$a2$v2$
+}
+
+Test-Case: output
+Description:  Test every character that is not part of an interpolation is interpreted as a literal character.
```
Comment (MaryamZi) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:222]: Add more tests with new lines, \n, \t, etc.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+Test-Case: output
+Description: Test string-template-expr is evaluated by evaluating the expression in each interpolation in the order in
+which they occur.
+Labels: string, function-defn, function-call-expr, string-template-expr, int, boolean
+
+function init() {
+    string a1 = "a1";
+    string b1 = "b1";
+    string temp1 = string `${a1} ${b1}`;
+    io:println(temp1); // @output a1 b1
```
Comment (MaryamZi) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:279]: How does this test the order in which the interpolations are evaluated?

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+    int a2 = 12;
+    string temp2 = string `${a2} ${b1+ a1}`;
```
Comment (MaryamZi) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:281]:
```suggestion
    string temp2 = string `${a2} ${b1 + a1}`;
```

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+    io:println(temp1); // @output `
+}
+
+Test-Case: error
+Description: Test interpolation without expression
+Labels: string, function-defn, string-template-expr
+
+function errorFunction() {
+    string _ = string `${}`; // @error missing expression
+}
```
Comment (MaryamZi) [inline conformance/lang/expressions/string-template-expr/string_template_expr.balt:323]: Please add a test or two to explicitly check "... and converting the result of the each evaluation to a string using the [ToString](https://ballerina.io/spec/lang/master/#ToString) abstract operation with the direct style".

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 5 above)
Comment (chiranSachintha) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt]: No. right? we didn't use the `boolean` type

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 8 above)
Comment (chiranSachintha) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt:242]: This is `Fail-Issue`

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 9 above)
Comment (chiranSachintha) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt:253]: These two functions are added in `Fail-Issue`

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 10 above)
Comment (chiranSachintha) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt]: Removed from here.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 11 above)
Comment (chiranSachintha) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt]: Updated the description. Please check.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt (hunk 12 above)
Comment (chiranSachintha) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt]: Removed.

File: conformance/lang/expressions/string-template-expr/string_template_expr.balt [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+    io:println(temp2); // @output 12
+    int a3 = 13;
+    string temp3 = string `${a2 + a3}`;
+    io:println(temp3); // @output 25
+    string temp4 = string `${"a2"}`;
+    io:println(temp4); // @output a2
+}
+
+Test-Case: error
+Description: Test whether string-template-expr interpolates the results of evaluating expressions into a literal string.
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply conformance/lang/expressions/string-template-expr/string_template_expr.balt]: Change the description.

**Pull request closed without merging**